journal: add monotonic_timestamp() and other fixes

jlebon opened the pull request:
See individual commit messages for details, the primary one being the second commit:

This function is not only useful for the timestamp itself, but also for
retrieving the boot ID of the current journal entry. IIUC, this is *not*
equivalent to the `_BOOT_ID` entry field, which may not be matched
properly in some cases. For more details, see:
systemd/systemd@dc00966

Implementation-wise, there's nothing surprising here. It's a
straightforward wrapper of the underlying API.  Also implement
`From<ffi::id128::sd_id128_t>` for `Id128` to convert to the wrapper
type from the ffi type and `PartialEq` to help testing (but which I also
need for my use case).

jlebon commented on tests/journal.rs:
```diff
@@ -55,6 +56,11 @@ fn ts() {
+    let (u, entry_boot_id) = j.monotonic_timestamp().unwrap();
+    assert!(u > 0);
+    let boot_id = id128::Id128::from_boot().unwrap();
+    assert!(boot_id == entry_boot_id);
```
Aside: I spent a while trying to get this to pass locally. It turned out to be due to the fact I was running it from a container. For reasons I still don't fully understand, the boot ID from the journal entry was different there even though I had full access to the host journals (and e.g. `journalctl --list-boots` in the container gave the right boot IDs).

I ended up testing this in a VM and it passed fine there!

cgwalters replied:
Did you make `/etc/machine-id` the same on the host and container?

jlebon replied:
Yup! That was a requirement to get `journalctl` to show any logs at all in the container.

lucab commented on src/journal.rs:
```diff
@@ -325,13 +326,25 @@ impl Journal {
+    /// Returns monotonic timestamp and boot ID at which current journal entry is recorded.
+    pub fn monotonic_timestamp(&self) -> Result<(u64, Id128)> {
+        let mut monotonic_timestamp_us: u64 = 0;
+        let mut id: sd_id128_t = unsafe { uninitialized() };
```
I think this is unidiomatic and you can't actually prove that the uninitialized memory is not leaked out (that depends on `ffi::sd_journal_get_monotonic_usec` internals).
I'd suggest to implement the equivalent of `SD_ID128_NULL` (see `sd-id128(3)`) on `Id128` and use that instead for the initialization here.

jlebon replied:
This pattern is lifted from `id128.rs`, though I agree we should use something like `SD_ID128_NULL`. Mind if I do that in a follow-up?

> that depends on `ffi::sd_journal_get_monotonic_usec` internals

Interesting, can you expand on that? If the call fails, then the value gets dropped right? Also note it's allocated on the stack.

lucab commented on src/id128.rs:
```diff
@@ -46,3 +46,21 @@ impl Id128 {
     }
 }
+
+impl From<ffi::id128::sd_id128_t> for Id128 {
```
This exposes a type from a different crate in the public API (which is generally problematic for semver). Also, I think that the idea is not to expose unsafe FFI bits in the safe high-level API.
I think a better approach would be to keep this as a private method. Or to implement fallible conversion from the two slice forms and with UUID semantic validation.

jlebon replied:
I've changed this to a crate-level `from_ffi()` instead.

lucab commented on src/journal.rs:
```diff
@@ -325,13 +326,25 @@ impl Journal {
     }
 
+    /// Returns monotonic timestamp and boot ID at which current journal entry is recorded.
+    pub fn monotonic_timestamp(&self) -> Result<(u64, Id128)> {
```
It looks like this is missing an input parameter compared to the C libary, i.e. the `If the boot ID parameter is passed NULL` case. Would that better be a separate method or a `current_boot_only: bool` parameter here? (I'm for the latter)

jlebon replied:
Yeah, good point. I made this a separate function instead because otherwise it'd require a separate `sd_id128_get_boot()` call afterwards in the `current_boot_only == true` case to still return the boot ID. Or I guess we could've changed the return type to something like `Result<(u64, Option<Id128>)`, but that just looks painful to use.

jlebon commented:
Thanks @lucab for reviewing! I've changed the `From<sd_id128_t>` to a crate-level `from_ffi()` instead. One hesitation I had was whether we were OK relying on `pub(crate)`; as @lucab mentioned, it was released in [1.18](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1180-2017-06-08) and it looks like e.g. Debian is at [1.24](https://packages.debian.org/stable/rust/) so I think we're fine in that regard (or is there a minimum version lower than this that we're targeting?).

In the process of doing these fixups, I've also finally resolved the issue I was hitting in https://github.com/jmesmon/rust-systemd/pull/54#discussion_r223141309. It has nothing to do with running in a container at all. The issue was that we were calling `sd_journal_next()` instead of `sd_journal_previous()` in `seek()` even in the case where we were seeking to the tail of the journal. See the second commit here for details.

Let me know if you'd rather I split this into multiple PRs!

jmesmon commented:
bors r+

lucab commented on src/journal.rs:
```diff
@@ -309,7 +313,11 @@ impl Journal {
             // Cursor may need to be re-aligned on a real entry first.
-            sd_try!(ffi::sd_journal_next(self.j));
+            if tail {
+                sd_try!(ffi::sd_journal_previous(self.j));
```
@jlebon did you run a `cargo test` after this change? On my local systemd (v239) this broke the tests:
```
test test_simple_match ... FAILED

---- test_simple_match stdout ----
thread 'test_simple_match' panicked at 'assertion failed: r.is_some()', tests/journal.rs:100:5
```

(and I think that travis does not have a recent enough journald running to check this)

lucab replied:
That's indeed the case. For some reason this just made the test quite reliably losing the race.
